src/vsr/replica.zig

```diff
@@ -3021,6 +3043,51 @@ pub fn ReplicaType(
             }
         }
 
+        fn on_grid_scrub_timeout(self: *Self) void {
+            assert(self.grid_scrub_timeout.ticking);
+            self.grid_scrub_timeout.reset();
+
+            if (!self.state_machine_opened) return;
+            if (self.syncing != .idle) return;
+            if (self.sync_tables != null) return;
+            assert(self.grid.callback != .cancel);
+
+            self.grid_scrub_timeout.after = std.math.clamp(
+                @divFloor(
+                    constants.grid_scrubber_cycle_ticks,
+                    @max(1, self.grid.free_set.count_acquired()),
+                ),
+                constants.grid_scrubber_interval_ticks_min,
+                constants.grid_scrubber_interval_ticks_max,
+            );
+
+            while (self.grid.blocks_missing.enqueue_blocks_available() > 0) {
+                const reclaim = self.grid_scrubber.read_reclaim() orelse break;
+                defer self.grid_scrubber_reads.release(reclaim.read);
+
+                if (reclaim.repair) |fault| {
+                    assert(!self.grid.free_set.is_free(fault.block_address));
+
+                    log.debug("{}: on_grid_scrub_timeout: fault found: " ++
+                        "block_address={} block_checksum={x:0>32} block_type={s}", .{
+                        self.replica,
+                        fault.block_address,
+                        fault.block_checksum,
+                        @tagName(fault.block_type),
+                    });
+
+                    self.grid.blocks_missing.enqueue_block(
+                        fault.block_address,
+                        fault.block_checksum,
+                    );
+                }
+            }
+
+            if (self.grid_scrubber_reads.acquire()) |read| {
```

Comment:
If we allow several scrubber reads, shouldn't this be a `while`? That is, try to enque many concurrent reads if we can?

Reply:
:+1: As part of this the timeout interval calculation needs to account for the number of concurrent scrub reads:

```
            self.grid_scrub_timeout.after = std.math.clamp(
                @divFloor(
                    constants.grid_scrubber_cycle_ticks,
                    @max(1, self.grid.free_set.count_acquired()),
                ) * constants.grid_scrubber_reads_max,
                constants.grid_scrubber_interval_ticks_min,
                constants.grid_scrubber_interval_ticks_max,
            );
```